OctopusDeploy/OctoPack
# [do not merge] Support NuGet 3.x and project.json

## What is this PullRequest about
When one is trying to use `Octopack` with NuGet 3.x and `project.json` file current setup will not work.

Based on NuGet 3.3 [documentation](https://docs.nuget.org/create/creating-and-publishing-a-package)

>Install.ps1 runs when a package is installed in a project.
>
>[NuGet 3.x] This script **will not be executed in projects managed by project.json**

However: 

>**Import MSBuild targets and props files into project**
>
>A new convention has been added to the structure of NuGet packages. 
As a peer to `\lib`, `\content`, and `\tools`, you can now include a `\build` folder in your package. 
Under this folder, you can place two files with fixed names, `{packageid}.targets` or `{packageid}.props`. 
These two files can be either directly under `\build` or under framework-specific folders just like the other folders. The rule for picking the best-matched framework folder is exactly the same as in those.

Technically there is no need for `Install.ps1`. 

>When NuGet installs a package with `\build` files, it will add an MSBuild element in the project file pointing 
to the `.targets` and `.props` files. The `.props` file is added at the top, whereas the `.targets` file is added to the bottom.

That will only work when one is using NuGet 2.5 or above. 
I deleted `Install.ps1` and `Uninstall.ps1` but they can be reverted to point to new `\build` folder for backward compatibility. 
I'm not sure if that's a good idea, though.

## Plan

I can see the migration as a few steps process:

* Modify `build.ps1` to `tools` content into `\build` folder
* Modify `.nuspec` file to copy `build\*` content into `build` target
* Delete `Install.ps1` and `Uninstall.ps1` files

## Implementation

`build.ps1`
```diff
-    mkdir .\build\tools
-    dir -recurse .\source\OctoPack.Tasks\bin\$configuration | copy -destination build\tools -Force
-    dir -recurse .\source\tools | copy -destination build\tools -Force
+    mkdir .\build\build
+    dir -recurse .\source\OctoPack.Tasks\bin\$configuration | copy -destination build\build -Force
+    dir -recurse .\source\tools | copy -destination build\build -Force
```

`Octopack.nuspec`
```diff
-    <file src="tools\*" target="tools" />
+    <file src="build\*" target="build" />
```

There was already a similar [PR](https://github.com/OctopusDeploy/OctoPack/pull/64), but it had a couple of formatting issues and seems to be abandoned now.

## How to test

Well, that's a tricky question. Honestly, I can't come up with a nice integration test. And I see the test as a redundant here.
It will not test `OctoPack` but it will test `NuGet`. And it should already be covered by NuGet [tests](https://github.com/NuGet/NuGet.Client/tree/dev/test).

What I did: 
* Built package with `build.ps1`
* Deployed it to the temporary NuGet feed
* Installed it via NuGet 3 UI
* Executed MSBuild command in the same folder

``"c:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe" /p:RunOctoPack=true``

and got the desired message :)

> OctoPack: OctoPack successful

Please let me know if there is anything that I can improve.

## Progress

- [x] Sync with remote, resolve conflicts
- [ ] Make sure that NuGet 2.xx is still supported
